### PR TITLE
Node 20 fixes

### PR DESCRIPTION
**Pull request by kanongil**
This fixes the test suite for node v18, v20 and v21 on my mac.

### CODE REVIEW

**Comment by kanongil** [comment]
Hmm, strange that the `macOS` runners fail, when they work locally (on M1 macOS 14.1.1).

Looking at the failing tests, it seems that `request.active()` never returns `false`.
The "*Request  active() exits handler early when request is no longer active*" test case seems central.

The "*Request _postCycle() skips onPreResponse when validation terminates request*" test does not directly use `request.active()` but it might still rely on the internal version of this.

I don't have a macOS 12 machine to test on, so I'm not really able to debug further.

Even stranger, the tests [passes on node 18](https://github.com/hapijs/hapi/actions/runs/6800372695/job/18488761001?pr=4472), but [fails again on node 20](https://github.com/hapijs/hapi/actions/runs/6800372695/job/18488760694?pr=4472#step:5:39).

**Comment by Nargonath** [comment]
Looks like we have some flakiness at play here. I restarted the 3 failed macos jobs and now macos node@14 passes whereas it didn't before. This might explain why it works on your machine @kanongil but not in the CI.

File: lib/transmit.js [hunk 1]
```diff
@@ -284,9 +284,7 @@ internals.end = function (env, event, err) {
     env.team = null;
 
     if (request.raw.res.writableEnded) {
-        if (!event) {
-            request.info.responded = Date.now();
-        }
+        request.info.responded = Date.now();
```
Comment (Nargonath) [inline lib/transmit.js:287]: Why did you remove the `!event` check?

File: test/transmit.js [hunk 2]
```diff
@@ -428,7 +428,7 @@ describe('transmission', () => {
             const log = server.events.once('response');
             const client = Net.connect(server.info.port, () => {
 
-                client.write('GET / HTTP/1.1\r\naccept-encoding: gzip\r\n\r\n');
+                client.write('GET / HTTP/1.1\r\nHost: host\r\naccept-encoding: gzip\r\n\r\n');
```
Comment (Nargonath) [inline test/transmit.js:431]: Why did you add the `Host` header?

**Review by Nargonath — APPROVED** [review]
code LGTM, I had some questions for my own curiosity.

File: lib/transmit.js (hunk 1 above)
Comment (kanongil) [reply lib/transmit.js:287]: Because the early `req.destroy()` call removes the only test case that has a non-`null` event, making it fail coverage. I don't think there are any other ways to end up with `res.writableEnded` along with an event, and if there is such a corner case, the damage will be negligible.

File: test/transmit.js (hunk 2 above)
Comment (kanongil) [reply test/transmit.js:431]: Because it is required to be a valid HTTP/1.1 request, and the node v20+ parser is more pedantic, no longer generating requests from this.

**Comment by kanongil** [comment]
> Looks like we have some flakiness at play here. I restarted the 3 failed macos jobs and now macos node@14 passes whereas it didn't before. This might explain why it works on your machine @kanongil but not in the CI.

Ah, it seems that it only fails on runners using macOS 12.6.9, while runners on macOS 12.7 works. I guess not all `macos-latest` runners are identical…

I have been doing some experiments, and the tests seem to pass when using the `macos-11` or `macos-13` (beta) runners, so that might be a way to fix this for now.

File: lib/transmit.js (hunk 1 above)
Comment (Nargonath) [reply lib/transmit.js:287]: Alright, thanks for the explanation.

File: test/transmit.js (hunk 2 above)
Comment (Nargonath) [reply test/transmit.js:431]: Ok, thanks.

**Comment by Nargonath** [comment]
Thanks for the investigation on the CI. We probably shouldn't block this PR for this and can handle it in another PR.
